### PR TITLE
Custom status feature

### PR DESCRIPTION
**Pull request by chetanyakan**
#### Summary
- Create custom status components.
    - Custom status option in status modal
    - Custom status modal
- Actions to set/unset status.

#### Related Pull Requests
- https://github.com/brightscout-alpha/mattermost-redux/pull/3

### CODE REVIEW

File: actions/views/custom_status.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        userProps.customStatusInitialisationState = JSON.stringify(initialState);
+        user.props = userProps;
+        await dispatch(updateMe(user));
+    };
+}
+
+export function clearCustomStatusInitialisationState() {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
+        const user = {...getCurrentUser(getState())};
+        const userProps = {...user.props};
```
Comment (harshilsharma63) [inline actions/views/custom_status.ts]: Any reason we can do `user.props.customStatusInitialisationState = ''`? Is it to forcefully create a new object?

File: components/custom_status/custom_status.scss [hunk 2]
```diff
@@ -0,0 +1,181 @@
+.StatusModal {
+    width: 600px !important;
```
Comment (harshilsharma63) [inline components/custom_status/custom_status.scss]: Any way to avoid `!important` here?

File: components/status_dropdown/status_dropdown.jsx [hunk 3]
```diff
@@ -116,10 +171,101 @@ export default class StatusDropdown extends React.PureComponent {
         const setAway = needsConfirm ? () => this.showStatusChangeConfirmation('away') : this.setAway;
         const setOffline = needsConfirm ? () => this.showStatusChangeConfirmation('offline') : this.setOffline;
 
+        let customStatusComponent;
+        if (this.props.isCustomStatusEnabled) {
+            let customStatusText = localizeMessage('status_dropdown.set_custom', 'Set a Custom Status');
```
Comment (harshilsharma63) [inline components/status_dropdown/status_dropdown.jsx]: We can move this code block in a separate function to make if block cleaner.

File: utils/post_utils.jsx [hunk 4]
```diff
@@ -486,3 +486,56 @@ export function makeGetReplyCount() {
...
+    if (!(post && previousPost)) {
+        return false;
+    }
+    return post.user_id === previousPost.user_id && // The post is by the same user
+        post.create_at - previousPost.create_at <= Posts.POST_COLLAPSE_TIMEOUT && // And was within a short time period
+        !(post.props && post.props.from_webhook) && !(previousPost.props && previousPost.props.from_webhook) && // And neither is from a webhook
+        !isSystemMessage(post) && !isSystemMessage(previousPost); // And neither is a system message
+}
+
+export function getPostFromId(state, id) {
```
Comment (harshilsharma63) [inline utils/post_utils.jsx]: This function name sounds like a generic function for fetching post by ID, but it does more than that. Can this be renamed?

**Review by harshilsharma63 — CHANGES_REQUESTED** [review]

File: components/custom_status/custom_status.scss (hunk 2 above)
Comment (manojmalik20) [reply components/custom_status/custom_status.scss]: This is to override the styles of another class. Added this comment in the code to explain.

**Comment by chetanyakan** [comment]
Closing as merged to MM via another PR.

**Pull request closed without merging**